### ISSUE TITLE
compilation failed on arm64

### ISSUE THREAD

**Issue by gfrankliu** [grpc/grpc#35052]
### What version of gRPC and what language are you using?

Release v1.59.3, cpp

### What operating system (Linux, Windows,...) and version?

Linux, Ubuntu 22.04

`Linux ubuntu-arm64 5.15.0-88-generic #98-Ubuntu SMP Mon Oct 2 15:23:16 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux`

### What runtime / compiler are you using (e.g. python version or version of gcc)

```
gcc --version
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

### What did you do?

Tried to compiled a test program provided as part of grpc on my arm64 Ubuntu to see if the OS is supported.

### What did you expect to see?

I expect to see the bundled test program can be compiled successfully.

### What did you see instead?

```
bazel build //test/cpp/util:channelz_sampler 

...
At global scope:
cc1plus: note: unrecognized command-line option '-Wno-typedef-redefinition' may have been intended to silence earlier diagnostics
cc1plus: all warnings being treated as errors
Target //test/cpp/util:channelz_sampler failed to build

```

**Comment by gfrankliu** [comment]
Tried Release v1.58.2 and got different failure:

```
Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
In file included from /usr/include/string.h:535,
                 from external/upb/upb/collections/array_internal.h:31,
                 from bazel-out/aarch64-opt-exec-2B5CBBC6/bin/external/com_google_protobuf/src/google/protobuf/descriptor.upb.h:12,
                 from external/upb/upbc/protoc-gen-upbdefs.cc:28:
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void _upb_MiniTable_CopyFieldData(void*, const void*, const upb_MiniTableField*)' at external/upb/upb/message/accessors_internal.h:119:13,
    inlined from 'void _upb_Message_GetNonExtensionField(const upb_Message*, const upb_MiniTableField*, const void*, void*)' at external/upb/upb/message/accessors_internal.h:210:33,
    inlined from 'const upb_Array* upb_Message_GetArray(const upb_Message*, const upb_MiniTableField*)' at external/upb/upb/message/accessors.h:297:36:
/usr/include/aarch64-linux-gnu/bits/string_fortified.h:29:33: error: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' writing 16 bytes into a region of size 8 overflows the destination [-Werror=stringop-overflow=]
   29 |   return __builtin___memcpy_chk (__dest, __src, __len,
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
   30 |                                  __glibc_objsize0 (__dest));
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
In function 'void* memcpy(void*, const void*, size_t)',
    inlined from 'void _upb_MiniTable_CopyFieldData(void*, const void*, const upb_MiniTableField*)' at external/upb/upb/message/accessors_internal.h:119:13,
    inlined from 'void _upb_Message_GetNonExtensionField(const upb_Message*, const upb_MiniTableField*, const void*, void*)' at external/upb/upb/message/accessors_internal.h:213:31,
    inlined from 'const upb_Array* upb_Message_GetArray(const upb_Message*, const upb_MiniTableField*)' at external/upb/upb/message/accessors.h:297:36:
/usr/include/aarch64-linux-gnu/bits/string_fortified.h:29:33: error: 'void* __builtin___memcpy_chk(void*, const void*, long unsigned int, long unsigned int)' writing 16 bytes into a region of size 8 overflows the destination [-Werror=stringop-overflow=]
   29 |   return __builtin___memcpy_chk (__dest, __src, __len,
      |          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~
   30 |                                  __glibc_objsize0 (__dest));
      |                                  ~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
Target //test/cpp/util:channelz_sampler failed to build
```

**Comment by gfrankliu** [comment]
I tried a few old versions: v1.56.1, v.1.56.3, v.1.55.4 are failed with the same error as above v1.58.2

v.1.54.3 works!

I also tried the latest from "master". It gave the same error as v1.59.3, but compilation succeeded:

```
cc1plus: note: unrecognized command-line option '-Wno-typedef-redefinition' may have been intended to silence earlier diagnostics
Target //test/cpp/util:channelz_sampler up-to-date:
  bazel-bin/test/cpp/util/channelz_sampler
INFO: Elapsed time: 173.604s, Critical Path: 23.92s
INFO: 2909 processes: 394 internal, 2515 linux-sandbox.
INFO: Build completed successfully, 2909 total actions
```

**Comment by gfrankliu** [comment]
It turned out that the reason I only saw the issue on my Ubuntu arm64 was because I didn't install "clang".  On my x86_64 machine, if I were to remove "clang", I would see the same issue. Is "clang" a mandatory requirement for grpc build?

BTW, is --client_env an official bazel command option? I checked the [bazel command-line reference](https://bazel.build/reference/command-line-reference) but couldn't find any mentioning of `--client_env`.
